### PR TITLE
Convert ChunkTransformer to GraphStage

### PR DESCRIPTION
**Pull request by poojadshende**

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala [hunk 1]
```diff
@@ -54,19 +57,34 @@ private object RenderSupport {
...
+        })
+
+        setHandler(out, new OutHandler {
+          override def onPull(): Unit = pull(in)
+        })
       }
+
+    override def onUpstreamFinish(): Unit = {
+      //terminationEmit(Iterator.single(defaultLastChunkBytes), out)
+      //completeStage()
```
Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala]: commented out?

**Comment by johanandren** [comment]
OK TO TEST

File: akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala [hunk 2]
```diff
@@ -54,19 +57,34 @@ private object RenderSupport {
...
+          override def onPush(): Unit = {
+            val chunk = grab(in)
+            val bytes = renderChunk(chunk)
+            if (chunk.isLastChunk) completeStage()
+          }
+        })
+
+        setHandler(out, new OutHandler {
+          override def onPull(): Unit = pull(in)
+        })
```
Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala]: The handlers can be implemented by the `GraphStageLogic` object itself to save two allocations on materialization (as we do in other built in stages).

File: akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala [hunk 3]
```diff
@@ -54,19 +57,34 @@ private object RenderSupport {
...
+    val in: Inlet[HttpEntity.ChunkStreamPart] = Inlet("ChunkTransformer.in")
+    val shape: FlowShape[HttpEntity.ChunkStreamPart, ByteString] = FlowShape.of(in, out)
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+      new GraphStageLogic(shape) {
+        setHandler(in, new InHandler {
+          override def onPush(): Unit = {
+            val chunk = grab(in)
+            val bytes = renderChunk(chunk)
+            if (chunk.isLastChunk) completeStage()
```
Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala]: `bytes` are never pushed

**Comment by johanandren** [comment]
This isn't ready for review, right?

File: akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala (hunk 1 above)
Comment (poojadshende) [inline akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala]: @johanandren. Thanks for feedback. But, what is replacement for terminationEmit in GraphStage.

Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala]: I think that would be 
```
emit(out, element)
completeStage()
```

Comment (poojadshende) [inline akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala]: I gives me a error 
```
not found: value emit
not found: value completeStage
```
But completeStage works fine in onPush()

Comment (poojadshende) [inline akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala]: It is working but I am not getting what should be the third argument for emit. I looked in GraphStage class for emit function and it takes three arguments. I used ``` DoNothing ``` as third argument but, it is gives me error 
``` not found: value DoNothing```

File: akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala [hunk 4]
```diff
@@ -54,19 +57,34 @@ private object RenderSupport {
...
+          if (chunk.isLastChunk) {
+            push(out, bytes)
+            completeStage()
+          } else push(out, bytes)
+        }
+
+        override def onPull(): Unit = pull(in)
+
+        override def onUpstreamFinish(): Unit = {
+          emitMultiple(out, Iterator.single(defaultLastChunkBytes))
```
Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala]: No need to create an iterator when you can just emit that single value using `emit(out, defaultLastChunkBytes)`

File: akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala [hunk 5]
```diff
@@ -54,19 +57,34 @@ private object RenderSupport {
   }
 
   object ChunkTransformer {
-    val flow = Flow[ChunkStreamPart].transform(() ⇒ new ChunkTransformer).named("renderChunks")
+    val chunkTransformer = Flow.fromGraph(new ChunkTransformer)
+    val flow = Flow[ChunkStreamPart].via(chunkTransformer).named("renderChunks")
```
Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala]: Not sure what's going on here, can't you just do like this?
```
val flow = Flow.fromGraph(new ChunkTransformer).named("renderChunks")
```

Comment (poojadshende) [inline akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala]: @johanandren I will make changes and commit.

File: akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala [hunk 6]
```diff
@@ -54,19 +57,35 @@ private object RenderSupport {
...
+    val in: Inlet[HttpEntity.ChunkStreamPart] = Inlet("ChunkTransformer.in")
+    val shape: FlowShape[HttpEntity.ChunkStreamPart, ByteString] = FlowShape.of(in, out)
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+      new GraphStageLogic(shape) with InHandler with OutHandler {
+        override def onPush(): Unit = {
+          val chunk = grab(in)
+          val bytes = renderChunk(chunk)
+          if (chunk.isLastChunk) {
+            push(out, bytes)
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala]: Move the `push` above the `if` so you can drop the `else` branch:
```scala
push(out, bytes)
if (chunk.isLastChunk) completeStage()

```

**Comment by poojadshende** [comment]
@2beaucoup I will make the change and commit.

**Comment by patriknw** [comment]
LGTM

**Comment by ktoso** [comment]
LGTM, thanks!

**Pull request merged**